src/main/java/sirius/biz/storage/layer3/VirtualFile.java

```diff
@@ -1267,6 +1270,124 @@ public boolean loadFromUrl(URL url, boolean force) throws IOException {
+     *         we use this.
+     *     </li>
+     *     <li>
+     *         Otherwise, we check the path in the URL. This the file there has an accepted file extension, we use this
+     *         path.
+     *     </li>
+     *     <li>
+     *         If the two attempts above fail, we emit a HEAD request and try to determine the filename/path by checking
+     *         the <tt>content-disposition</tt> header.
+     *     </li>
+     * </ol>
+     *
+     * @param url                   the URL which determines the filename/path as well as the source of the file to fetch
+     * @param force                 if set to <tt>true</tt> a download will be performed, even if we already downloaded the file
+     *                              before and no modification was detected.
+     * @param fileExtensionVerifier specifies which extensions are accepted. This should be used to prevent using
+     *                              ".php" or the like as effective file name.
+     * @return the file which has been resolved (and downloaded if necessarry)
+     * @throws IOException      in case of an IO error during the download
+     * @throws HandledException in case no effective filename can be detected
+     */
+    public VirtualFile resolveOrLoadChildFromURL(URL url, boolean force, Predicate<String> fileExtensionVerifier)
+            throws IOException {
+        String path = loadPathFromUrl(url, fileExtensionVerifier);
+        if (Strings.isEmpty(path)) {
+            String contentDisposition = loadContentDisposition(url);
+            if (fileExtensionVerifier.test(contentDisposition)) {
+                path = contentDisposition;
+            } else {
+                throw Exceptions.createHandled()
+                                .withNLSKey("VirtualFile.loadFromUrl.noValidPath")
+                                .set("url", url)
+                                .handle();
+            }
+        }
+        VirtualFile file = resolve(path);
+        file.loadFromUrl(url, force);
+        return file;
+    }
+
+    private String loadContentDisposition(URL url) throws IOException {
+        HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
+        urlConnection.connect();
+        String contentDisposition = urlConnection.getHeaderField(HttpHeaderNames.CONTENT_DISPOSITION.toString());
+        urlConnection.disconnect();
+        return contentDisposition;
+    }
+
+    @Nullable
+    private String loadPathFromUrl(URL url, Predicate<String> fileExtensionVerifier) {
+        QueryStringDecoder qsd = new QueryStringDecoder(url.toString(), StandardCharsets.UTF_8);
+        return qsd.parameters()
+                  .values()
+                  .stream()
+                  .flatMap(List::stream)
+                  .filter(fileExtensionVerifier)
+                  .findFirst()
+                  .orElseGet(() -> {
+                      if (fileExtensionVerifier.test(url.getPath())) {
```

Comment:
fileExtensionVerifier should only be supplied with the file extension, not the whole path - or wrap this function in a call chain via Files.getFileExtension (or whathever it is called)